### PR TITLE
MINOR: Clean up of TransactionManager and RecordAccumulator

### PR DESCRIPTION
**Pull request by ddrid**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ddrid** [comment]
Hi, @showuon @dajac, could you please take a look at this easy one?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java [hunk 1]
```diff
@@ -378,12 +378,12 @@ public int splitAndReenqueue(ProducerBatch bigBatch) {
...
     // IllegalStateException instead.
     private void insertInSequenceOrder(Deque<ProducerBatch> deque, ProducerBatch batch) {
-        // When we are requeing and have enabled idempotence, the reenqueued batch must always have a sequence.
+        // When we are re-enqueueing and have enabled idempotence, the re-enqueued batch must always have a sequence.
         if (batch.baseSequence() == RecordBatch.NO_SEQUENCE)
             throw new IllegalStateException("Trying to re-enqueue a batch which doesn't have a sequence even " +
                 "though idempotency is enabled.");
 
-        if (transactionManager.nextBatchBySequence(batch.topicPartition) == null)
+        if (!transactionManager.hasInflightBatches(batch.topicPartition))
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:386]: Nice cleanup. This way we don't have to return the first element in queue.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java [hunk 2]
```diff
@@ -113,16 +113,7 @@ private TopicPartitionEntry getPartition(TopicPartition topicPartition) {
...
-            if (ent == null) {
-                ent = new TopicPartitionEntry();
-                topicPartitions.put(topicPartition, ent);
-            }
-            return ent;
-        }
-
-        private void addPartition(TopicPartition topicPartition) {
-            this.topicPartitions.putIfAbsent(topicPartition, new TopicPartitionEntry());
+            return topicPartitions.putIfAbsent(topicPartition, new TopicPartitionEntry());
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java]: I don't think this refactor is correct. `putIfAbsent` will return **previous** value of the key. That is, if the current partition is null, after `getOrCreatePartition`, it'll return `null`, which is not what we want.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java [hunk 3]
```diff
@@ -790,7 +781,6 @@ private void adjustSequencesDueToFailedBatch(ProducerBatch batch) {
                 throw new IllegalStateException("Sequence number for batch with sequence " + inFlightBatch.baseSequence()
                         + " for partition " + batch.topicPartition + " is going to become negative: " + newSequence);
 
-            log.info("Resetting sequence number of batch with current sequence {} for partition {} to {}", inFlightBatch.baseSequence(), batch.topicPartition, newSequence);
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:793]: why should we remove this log?

**Review by showuon — COMMENTED** [review]
@ddrid , thanks for the PR. Left some comments.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java (hunk 2 above)
Comment (ddrid) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java]: Thanks for reminding me of that! I've confused it with `computeIfAbsent`. How about using `computeIfAbsent` as I modified?

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java (hunk 3 above)
Comment (ddrid) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:793]: Cause `inFlightBatch.resetProducerState` will be called immediately. I think this log is duplicate in `resetProducerState` method.

**Comment by ddrid** [comment]
Hi @showuon, could you take another look if you have time? thanks

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java (hunk 3 above)
Comment (showuon) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:793]: Aha, nice catch!

**Review by showuon — APPROVED** [review]
LGTM!

**Comment by showuon** [comment]
Failed tests are unrelated:
```
Build / ARM / kafka.network.ConnectionQuotasTest.testListenerConnectionRateLimitWhenActualRateAboveLimit()
```

**Pull request merged**